xds/internal/xdsclient/tests/ads_stream_flow_control_test.go

```diff
@@ -0,0 +1,623 @@
+func (s) TestADSFlowControl_ResourceErrors(t *testing.T) {
+	ctx, cancel := context.WithTimeout(context.Background(), defaultTestTimeout)
+	defer cancel()
+
+	// Override the ADS stream creation.
+	adsStreamCh := overrideADSStreamCreation(t)
+
+	// Start an xDS management server.
+	mgmtServer := e2e.StartManagementServer(t, e2e.ManagementServerOptions{})
+
+	// Create bootstrap configuration pointing to the above management server.
+	nodeID := uuid.New().String()
+	bc := e2e.DefaultBootstrapContents(t, nodeID, mgmtServer.Address)
+	testutils.CreateBootstrapFileForTesting(t, bc)
+
+	// Create an xDS client with the above bootstrap contents.
+	client := createXDSClient(t, bc)
+
+	// Configure a watcher for a listener resource.
+	const listenerResourceName = "test-listener-resource"
+	watcher := newBLockingListenerWatcher()
+	cancel = xdsresource.WatchListener(client, listenerResourceName, watcher)
+	defer cancel()
+
+	// Wait for the wrapped ADS stream to be created.
+	var adsStream *wrappedADSStream
+	select {
+	case adsStream = <-adsStreamCh:
+	case <-ctx.Done():
+		t.Fatalf("Timed out waiting for ADS stream to be created")
+	}
+
+	// Configure the management server to return a single listener resource
+	// which is expected to be NACKed by the client.
+	resources := e2e.UpdateOptions{
+		NodeID:         nodeID,
+		Listeners:      []*v3listenerpb.Listener{badListenerResource(t, listenerResourceName)},
+		SkipValidation: true,
+	}
+	if err := mgmtServer.Update(ctx, resources); err != nil {
+		t.Fatalf("Failed to update management server with resources: %v, err: %v", resources, err)
+	}
+
+	// Ensure that there is a read on the stream.
+	select {
+	case <-adsStream.recvCh:
+	case <-ctx.Done():
+		t.Fatalf("Timed out waiting for ADS stream to be read from")
+	}
+
+	// Wait for the error to reach the watcher.
+	select {
+	case <-watcher.errorCh:
+	case <-ctx.Done():
+		t.Fatalf("Timed out waiting for error to reach watcher")
+	}
+
+	// Wait for a short duration and ensure that there is no read on the stream.
+	select {
+	case <-adsStream.recvCh:
+		t.Fatal("Recv() called on the ADS stream before all watchers have processed the previous update")
+	case <-time.After(defaultTestShortTimeout):
+	}
+
+	// Unblock one watcher.
+	done := <-watcher.doneNotifierCh
+	done.OnDone()
+
+	// Ensure that there is a read on the stream, now that the previous error
+	// has been consumed by the watcher.
+	select {
+	case <-adsStream.recvCh:
+	case <-ctx.Done():
+		t.Fatalf("Timed out waiting for Recv() to be called on the ADS stream after all watchers have processed the previous update")
+	}
+}
+
+// Test ADS stream flow control with a single resource that deleted from the
+// management server and therefore the watcher's OnResourceDoesNotExist()
+// callback is expected to be invoked. Verifies that no further reads are
+// attempted until the callback is completely handled by the watcher.
+func (s) TestADSFlowControl_ResourceDoesNotExist(t *testing.T) {
```

Comment:
This seems similar to the test above albeit a few different operations. Can we refactor the two into a t-test? There seems to be a lot of shared code.

Reply:
Yes, there is about 30 lines of common code at the start of these two tests, but writing a table driven test to cater to these two scenarios would make the logic inside the table complicated. As you can see for the error case, we simply configure a resource on the management server which is expected to NACKed by the xDS client. But for the resource-not-found case, we initially need to have a resource which is expected to be ACKed by the xDS client and then we remove it from the management server triggering the resource-not-found error.

My preference is to keep tests simpler to read and maintain, over tests that are shorter and share more code, but end up being complicated.

Let me know your thoughts. Thanks.

Reply:
Alright reading the test bodies I guess the operations are different enough. I guess refactoring into setup would be hard across tests since variable listeners and routes.